DNX on Linux bring-up.

ellismg opened the pull request:
This seems like a good checkpoint.

The following changes make DNX cross platform and add another second level unamanged loader which can load CoreCLR on Linux and jump into it.  These components are not part of the official build yet, so instead Makefiles are provided to build them on an ad-hoc basis.

With these changes, it's possible to run a pre-compiled console hello world application, using CoreCLR on Linux, via dnx <path-to-hello-world.dll>.

The next step is to get HelloWorld running when compiled dynamically via Roslyn.

There are still some TODOs in the code that will get addressed later, for now this is enough to start getting things working.

msotclas commented:
Hi __@ellismg__, I'm your friendly neighborhood Microsoft Open Technologies, Inc. Pull Request Bot (You can call me MSOTBOT). Thanks for your contribution!
    <p>
        It looks like you're a Microsoft contributor (Matt Ellis). If you're full-time, we DON'T require a Contribution License Agreement. If you are a vendor, please DO sign the electronic Contribution License Agreement. It will take 2 minutes and there's no faxing! https://cla2.msopentech.com.
    </p>

TTYL, MSOTBOT;

BrennanConroy commented on src/dnx.coreclr.unix/dnx.coreclr.cpp:
```diff
@@ -0,0 +1,227 @@
+        return false;
+    }
+
+    trustedPlatformAssemblies = std::string();
```
.clear() instead

borgdylan commented:
Please do not remove mono support for now. CoreCLR is not ready for 32bit on Linux and EF7 is not mature enough for database stuff as of now.

davidfowl commented:
Nobody said we're removing support for mono.

davidfowl commented on src/dnx.coreclr.unix/tpa.cpp:
```diff
@@ -0,0 +1,105 @@
+// TODO: Unify this file with the one generate at build time for
```
I was about to comment and saw this :smile:

rpetrich commented on src/dnx.coreclr.unix/tpa.cpp:
```diff
@@ -0,0 +1,105 @@
+{
+    for (size_t idx = 0; idx < count; ++idx)
+    {
+        delete[] values[idx];
```
Should perhaps use `free` instead of `delete[]`

erkinalp replied:
rpetrich: `malloc` goes with `free`, `new` goes with `delete` and `delete[]`. They have different free lists and you will get a double free if you mix on the same object.

mmdriley replied:
Each of these values was allocated with strdup(), which is spec'd as allocating memory that must be deleted with free().
http://pubs.opengroup.org/onlinepubs/009695399/functions/strdup.html

Probably best we just get rid of the naked pointers altogether and use std::strings.
Best thing here is probably to use real strings and avoid the

ellismg replied:
I'll talk with the larger team about this.  I don't want to start making a bunch of changes to this file, since ideally we can unify it with the existing version.

davidfowl replied:
agreed

mmdriley replied:
erm, agreed with which part? the current code is incorrect, so it should not be checked in.

troydai replied:
Should have used `free`. I'll go update the build scripts to generate correct result.

ellismg replied:
@troydai I want to edit these scripts anyway, so I can take care of it.  Do you also know why we do strdup at all?  No one ever seems to modify the strings.  Can we just use literal strings and then get out of the free game?

troydai replied:
@ellismg I don't think anything blocking us from doing this. When I wrote these codes only my cat (she stay by my side at the time) and me knows why I use `_strdup`. Now only my cat knows.   JK. I guess one of the reasons was trying to reuse the data between crossgened file names and original names. But end with hard coded all the codes because it helps to allow simplier build scripts.

ellismg replied:
Fixed this all up and updated the generators.

mmdriley commented on src/Microsoft.Framework.Runtime.Loader/LoadContext.cs:
```diff
@@ -60,7 +60,7 @@ public static void InitializeDefaultContext(LoadContext loadContext)
         {
             var directory = Path.GetDirectoryName(ilPath);
-            var arch = Environment.GetEnvironmentVariable("PROCESSOR_ARCHITECTURE");
+            var arch = IntPtr.Size == 4 ? "x86" : "AMD64";
```
What about ARM?

ellismg replied:
I asked @davidfowl about exactly this and he said they used this pattern elsewhere in the stack as well, so there's already debt.  If we actually end up needing to detect this, I would much prefer adding an API to the BCL and reading that vs environment variables.

davidfowl replied:
Yep, we'll cross that bridge when we get to it. We're aware of the limitation.

mmdriley commented on src/dnx.coreclr.unix/dnx.coreclr.cpp:
```diff
@@ -0,0 +1,227 @@
+#define S_OK 0;
+#define E_FAIL -1;
+
+typedef struct CALL_APPLICTION_MAIN_DATA
```
'APPLICATION'

ellismg replied:
Fixed.

mmdriley commented on src/dnx.coreclr.unix/dnx.coreclr.cpp:
```diff
@@ -0,0 +1,227 @@
+// Windows types used by the ExecuteAssembly function
+typedef int32_t HRESULT;
+typedef const char* LPCSTR;
+typedef unsigned int DWORD;
```
more appropriately uint32_t

erkinalp replied:
Yes, `typedef int32_t DWORD;` will be more appropriate

ellismg replied:
DWORD is unsigned.  Did you mean uint32_t?

ellismg replied:
typedef unsigned int DWORD -> typedef uint32_t DWORD

ellismg commented:
It would appear that I included the DNX binary itself in these commits.  That was unintentional and I will rebase to remove it.

mmdriley commented on src/dnx.coreclr.unix/dnx.coreclr.cpp:
```diff
@@ -0,0 +1,227 @@
+typedef int32_t HRESULT;
+typedef const char* LPCSTR;
+typedef unsigned int DWORD;
+#define S_OK 0;
```
danger! semicolon in macro definition

let's just use const

ellismg replied:
Fixed.

mmdriley commented on src/dnx.coreclr.unix/dnx.coreclr.cpp:
```diff
@@ -0,0 +1,227 @@
+                    LPCSTR entryPointMethodsName,
+                    DWORD* exitCode);
+
+bool GetTrustedPlatformAssembliesList(const std::string& szDirectory, bool bNative, std::string& trustedPlatformAssemblies)
```
seems like we should skip the Hungarian for szDirectory, at least. Maybe bNative too since it's a C++ bool.

ellismg replied:
Fixed.

mmdriley commented on src/dnx.coreclr.unix/dnx.coreclr.cpp:
```diff
@@ -0,0 +1,227 @@
+
+    if (coreClrEnvVar != nullptr)
+    {
+        runtimeDirectory = std::string("CORECLR_DIR");
```
can just assign the string literal.
runtimeDirectory = "CORECLR_DIR";

BrennanConroy replied:
I think you meant
runtimeDirectory = coreClrEnvVar;

ellismg replied:
Fixed.

mmdriley commented on src/dnx.coreclr.unix/dnx.coreclr.cpp:
```diff
@@ -0,0 +1,227 @@
+        ret = dlopen(coreClrDllPath.c_str(), RTLD_NOW | RTLD_GLOBAL);
+    }
+
+    if (ret == nullptr)
```
if (!ret)

ellismg replied:
Fixed all the places we did an explicit test against nullptr.

mmdriley commented on src/dnx.coreclr.unix/dnx.coreclr.cpp:
```diff
@@ -0,0 +1,227 @@
+bool GetTrustedPlatformAssembliesList(const std::string& szDirectory, bool bNative, std::string& trustedPlatformAssemblies)
+{
+    size_t cTpaAssemblyNames = 0;
+    char** ppszTpaAssemblyNames = NULL;
```
consistency, NULL vs. nullptr. Prefer the latter.

ellismg replied:
Old habits die hard :-(.

ellismg replied:
Fixed.

mmdriley commented on src/dnx.coreclr.unix/dnx.coreclr.cpp:
```diff
@@ -0,0 +1,227 @@
+            runtimeDirectory.assign(pathToBootstrapper);
+
+            size_t lastSlash = runtimeDirectory.rfind('/');
+            runtimeDirectory.erase(lastSlash);
```
What if no slash is found?

ellismg replied:
Fixed (in the case rfind returns std::string::npos we just don't continue).

mmdriley commented on src/dnx.coreclr.unix/dnx.coreclr.cpp:
```diff
@@ -0,0 +1,227 @@
+
+    if (data->runtimeDirectory != NULL)
+    {
+        runtimeDirectory = std::string(data->runtimeDirectory);
```
just assignment is fine here

ellismg replied:
Fixed.

mmdriley commented on src/dnx.coreclr.unix/dnx.coreclr.cpp:
```diff
@@ -0,0 +1,227 @@
+            fprintf(stderr, "Failed to find files in the coreclr directory\n");
+            hr = E_FAIL;
+            return hr;
+            // goto Finished;
```
?

ellismg replied:
The code here makes more sense if you look at how dnx.coreclr works in windows.  This call normally has the side effect of probing on disk to make sure everything on the TPA list is present.  The logic matches the existing windows probing behavior for the second level loader.

mmdriley replied:
More curious about the 'goto' to a nonexistent label.

ellismg replied:
Ahh, yes.  I will remove this line.  It's vestigial code from a previous attempt at structuring this file.  Thanks!

ellismg replied:
Fixed.

mmdriley commented on src/dnx.coreclr.unix/stdafx.h:
```diff
@@ -0,0 +1,13 @@
+#include <string>
+#include <unistd.h>
+
+typedef char* LPWSTR;
```
seems wrong?

ellismg replied:
I agree but this is a point in time issue.  This is needed because I wanted to share as much of tpa.[h|cpp] between windows and *nix.  I will fix this later by moving to TCHAR like elsewhere.

ellismg replied:
Fixed now instead, since I ended up touching tpa.cpp anyway.

mmdriley commented on src/dnx.coreclr.unix/tpa.cpp:
```diff
@@ -0,0 +1,105 @@
+
+    if (bNative)
+    {
+        pArray[0] = strdup("Internal.Runtime.Uri.ni.dll");
```
this may be a job for vector<string> and std::initializer_list.
e.g.
std::vector<string> names = {
  "a!",
  "b!",
  // ...
};

ellismg replied:
I'll defer this feedback.  I want to unify this file with the existing generated version (but I will need to do a LPWSTR -> LPTSTR conversion first). Because of that, I don't want to make major changes.

mmdriley commented on src/dnx.coreclr.unix/tpa.h:
```diff
@@ -0,0 +1,4 @@
+#pragma once
+
+BOOL CreateTpaBase(LPWSTR** ppNames, size_t* pcNames, bool bNative);
```
this header file uses a lot of types it doesn't include.

ellismg replied:
The existing version of the file already does the same thing.  They assume that windows.h and friends will already be included.

mmdriley commented on src/dnx/dnx.cpp:
```diff
@@ -1,97 +1,107 @@
 {
-    int nIndex = wcsnlen_s(pszStr, MAX_PATH) - 1;
-    do
+    size_t nIndex = _tcsnlen(pszStr, MAX_PATH) - 1;
```
The use of MAX_PATH seems really odd here. There's nothing in the function name to suggest this is dealing with paths.

ellismg replied:
Fixed.

mmdriley commented on src/dnx/dnx.cpp:
```diff
@@ -1,97 +1,107 @@
+bool StringsEqual(LPCTSTR const pszStrA, LPCTSTR const pszStrB)
 {
-    return ::_wcsnicmp(pszStrA, pszStrB, MAX_PATH) == 0;
+    return ::_tcsnicmp(pszStrA, pszStrB, MAX_PATH) == 0;
```
Likewise on the use of MAX_PATH. Is this supposed to be "PathsEqual"?

ellismg replied:
Fixed.

BrennanConroy commented on src/dnx.coreclr.unix/dnx.coreclr.cpp:
```diff
@@ -0,0 +1,227 @@
+        trustedPlatformAssemblies.append(":");
+    }
+
+    if (ppszTpaAssemblyNames != NULL)
```
Seems odd to be doing a null check after the array has been used

ellismg replied:
I think the (unstated) invariant here is that if ppszTpaAssemblyNames is null, then cTpaAssemblyNames is zero.  This same weird pattern is already in dnx.coreclr.  I will add an assert.

BrennanConroy commented on src/dnx.coreclr.unix/dnx.coreclr.cpp:
```diff
@@ -0,0 +1,227 @@
+    // It looked like in some places in the CoreCLR it was hardcoded to just "CoreCLR".
+
+    char pathToBootstrapper[PATH_MAX];
+    ssize_t ret = readlink("/proc/self/exe", pathToBootstrapper, PATH_MAX - 1);
```
I liked the variable name pathLen you used earlier for similar code.

ellismg replied:
Fixed.

ellismg commented:
Thank you @BrennanConroy and @mmdriley.  I have pushed new history which addresses the feedback.  When it was appropriate, I just rewrote the commits that had problems instead of introducing new fixup commits.  Please take another look.

BrennanConroy commented on src/dnx.coreclr.unix/dnx.coreclr.cpp:
```diff
@@ -0,0 +1,229 @@
+    assert(ppszTpaAssemblyNames != nullptr || cTpaAssemblyNames == 0);
+
+    //TODO: The Windows version of this actaully ensures the files are present.  We just fail for native and assume MSIL is present
+    if (isNative)
```
nit: Move if check to above CreateTpaBase

ellismg replied:
I'm going to push back against this.  We need to remove the if (isNative) check in favor of doing the same thing for both native and the msil cases, which is looking on the file system to ensure all of the files from the TPA list are present.  For now, we just use the isNative check as a heuristic instead of actually probing the disk.  So long term the "should we return true or false from this method" decision is going to be made starting around L48.  If we move the code now, we'll just have to delete it later and it gets things further away from what we want, IMHO.

mmdriley commented on makefile.shade:
```diff
@@ -909,13 +911,8 @@ functions @{
         content.Add("}");
         content.Add(string.Empty);
-        content.Add("BOOL FreeTpaBase(const LPWSTR* values, const size_t count)");
+        content.Add("BOOL FreeTpaBase(const LPCTSTR* values, const size_t count)");
```
This no longer needs a "count" parameter.

ellismg replied:
Yes.  Fixed.

ellismg commented:
I don't believe there are any open issues here.  Anything left for me to do before we can merge this?

BrennanConroy commented:
I think it's good.

victorhurdugaci commented:
I was able to run it on Ubuntu :shipit:

ellismg commented:
Rebased on dev and pushed, 24dd1f8d
